design/ceph/object/store.md

```diff
@@ -712,3 +712,64 @@ For external CephCephObjectStores (i.e., when `spec.gateway.externalRgwEndpoints
+want access to rare and new features, Rook can allow users to set arbitrary Ceph configs in their
+CephObjectStore CR. RGW config reference: https://docs.ceph.com/en/reef/radosgw/config-ref/
+
+This is an advanced feature. It allows users to modify many untested aspects of the RGW and object
+store. Any user configs that break the object store should be a responsibility of the user.
+
+```yaml
+spec:
+  metadataPool: # ...
+  dataPool: # ...
+  gateway: # ...
+
+  cephConfig:
+    rgw_enable_apis: s3, s3website, swift, swift_auth, sts, iam, notifications # e.g. 'admin' removed
+    rgw_gc_max_concurrent_io: "4" # e.g, reduce garbage collection IO
+    debug_rgw: "20"
+```
+
+CephCluster has a similar configuration section called `config`. `config` may be somewhat too vague
+in the context of the CephObjectStore CR. `cephConfig` is a clearer naming for users and give
+context that Ceph configs are expected.
+
+Rook will apply user config values to the mon config store. Rook will not override its own mon
+config values set via other means. This will ensure that values controlled by Rook for other spec
+configurations or that are important for running in Kubernetes are not clobbered by user configs.
+This will risk some instances where users want to update a value that Rook currently sets.
+This tradeoff favors reliability over flexibility.
+
+Examples:
+    - Rook sets configs like `rgw_zone` which are critical for a functional CephObjectStore. Users
+    would not be able to override this value.
+    - Rook sets `rgw_dns_name` via the `hosting` config section. User inputs to `rgw_dns_name`
+    wouldn't take affect when `hosting` configs are set.
+
+If possible, Rook should make log when it overrides a user's `cephConfig` based on another config or
+internal decision. This may not be easily feasible when Rook applies CLI flags to the RGW
+deployment.
+
+Rook will apply each configuration under the client scope of the shared RGW client.
+i.e., `client.rook-ceph-rgw-<store-name>`. This will also help ensure that configs changed don't
+affect other Rook-managed Ceph daemons, critically mons, osds, and mds.
+
+Rook has the option to validate that each config is prefixed with `rgw_`. This would prevent users
+from modifying configs that are unrelated to RGW functionality. For example, users would be
+prevented from setting a config like `admin_socket` which might affect RGW's ability to run
+correctly in the container. As a downside, this would also mean that users are unable to set
+non-affecting configs like `debug_rgw`.
+[Ceph general configs](https://docs.ceph.com/en/reef/rados/configuration/general-config-ref/) that
+apply to all daemons are unlikely to be used by non-advanced users and probably unlikely to result
+in issues keeping RGWs running. Based on the perceived low risk, and the definite benefit for users
+especially for debugging the object store, Rook will not validate configs.
```

Comment:
@travisn thoughts?

Reply:
I'm not taking a position on if the keys should be filtered or not.  In terms of user visible behavior, it would be better to start off with a filter or allow list rather than try to add one later "because Hyrum's Law".

Reply:
This is a really good point! Thanks for that.

I will update my proposal to allow-list `rgw_*` configs and `debug_rgw*` configs. Those seem most important to me.

Reply:
@travisn Josh had a good alternate point here. I'm going to make an update to the design, and let's see what you think about it